File: packages/runtime/src/keys.ts
import type { NodeKeyTemplate } from "@acpus/core";
import type { NodeKeyDynamic } from "./types.js";

const DYNAMIC_SEGMENT = /^(item|lane|round|branch):/u;

export interface ParsedNodeKey {
  nodeKey: string;
  staticPath: string;
  staticSegments: string[];
  /** Collapsed dynamic dimensions (last-value-wins for repeated dimensions). */
  dynamic: NodeKeyDynamic;
  /**
   * Full frame-based parsing of dynamic dimensions. Each frame captures one
   * dynamic "scope" (e.g., from a fanout/parallel/loop parent). Repeated
   * dimensions appear in separate frames rather than being collapsed.
   */
  dynamicFrames: NodeKeyDynamic[];
}

type DynamicFrame = NodeKeyDynamic;

/**
 * Convert a NodeKeyTemplate + runtime dynamic context into a stable key string
 * for file naming and state lookup.
 *
 * Encoding: slash-separated path segments where dynamic dimensions use
 * `type:value` format. All values are filesystem-safe (no colons or
 * special chars in values — colons only appear as dimension separators).
 *
 * Examples:
 *   - Plain: "workflow/step-a"
 *   - Fanout item: "workflow/mapped/item:file-a"
 *   - Fanout with lane: "workflow/mapped/item:file-a/lane:0"
 *   - Loop round: "workflow/iterator/round:3"
 *   - Parallel branch: "workflow/parallel-group/branch:0"
 *   - Composite: "workflow/mapped/item:file-a/lane:0/branch:0/round:2"
 */
export function resolveNodeKey(
  template: NodeKeyTemplate,
  dynamic: NodeKeyDynamic = {}
): string {
  const parts: string[] = [template.nodePath];
  const frames = dynamic.frames && dynamic.frames.length > 0 ? dynamic.frames : [dynamic];

  // Always include dynamic dimensions when provided, regardless of template flags.
  // The template flags indicate the node *can* have that dimension, but at runtime
  // any ancestor that provides a dimension should be reflected in child keys too.
  for (const frame of frames) {
    if (frame.fanoutItemId !== undefined) {
      parts.push(`item:${sanitizeValue(String(frame.fanoutItemId))}`);
    }

    if (frame.laneId !== undefined) {
      parts.push(`lane:${sanitizeValue(String(frame.laneId))}`);
    }

    if (frame.parallelBranchId !== undefined) {
      parts.push(`branch:${sanitizeValue(String(frame.parallelBranchId))}`);
    }

    if (frame.loopRound !== undefined) {
      parts.push(`round:${String(frame.loopRound)}`);
    }
  }

  return parts.join("/");
}

/**
 * Parse a resolved Node Key into its static Node path and dynamic dimensions.
 *
 * Dynamic segments are slash-separated `type:value` segments matching the
 * DYNAMIC_SEGMENT pattern (`item:`, `lane:`, `branch:`, `round:`). Segments
 * that match are treated as dynamic; all others are static.
 *
 * **Ambiguity note**: A step ID like `branch:blue` would match DYNAMIC_SEGMENT
 * and be misparsed as dynamic. This is prevented at the compiler level:
 * authored step and branch IDs must match the safe ID pattern. The runtime
 * parsing relies on this compiler guarantee.
 */
export function parseNodeKey(nodeKey: string): ParsedNodeKey {
  const segments = nodeKey.split("/");
  const staticSegments: string[] = [];
  const dynamicSegments: string[] = [];

  for (const segment of segments) {
    if (DYNAMIC_SEGMENT.test(segment)) {
      dynamicSegments.push(segment);
    } else {
      staticSegments.push(segment);
    }
  }

  const dynamic: NodeKeyDynamic = {};
  for (const segment of dynamicSegments) {
    const [kind, ...rest] = segment.split(":");
    const value = rest.join(":");
    switch (kind) {
      case "item":
        dynamic.fanoutItemId = value;
        break;
      case "lane":
        dynamic.laneId = value;
        break;
      case "branch":
        dynamic.parallelBranchId = value;
        break;
      case "round":
        dynamic.loopRound = Number(value);
        break;
    }
  }

  const dynamicFrames = buildDynamicFrames(dynamicSegments);

  return {
    nodeKey,
    staticPath: staticSegments.join("/"),
    staticSegments,
    dynamic,
    dynamicFrames
  };
}

/**
 * Build dynamic frames from the already-identified dynamic segments.
 * Each time we encounter a new "item:" segment, we start a new frame
 * (since item: begins a new fanout scope). Other dimensions accumulate
 * into the current frame.
 */
function buildDynamicFrames(dynamicSegments: string[]): NodeKeyDynamic[] {
  const frames: NodeKeyDynamic[] = [];
  let current: NodeKeyDynamic = {};

  for (const segment of dynamicSegments) {
    const [kind, ...rest] = segment.split(":");
    const value = rest.join(":");

    // item starts a fanout frame; branch/round start their own composite frames.
    if ((kind === "item" || kind === "branch" || kind === "round") && !isEmptyDynamicScope(current)) {
      frames.push(current);
      current = {};
    }

    switch (kind) {
      case "item":
        current.fanoutItemId = value;
        break;
      case "lane":
        current.laneId = value;
        break;
      case "branch":
        current.parallelBranchId = value;
        break;
      case "round":
        current.loopRound = Number(value);
        break;
    }
  }

  if (!isEmptyDynamicScope(current)) {
    frames.push(current);
  }
  return frames;
}

/** Return the static IR nodePath represented by a resolved Node Key. */
export function staticNodePathFromKey(nodeKey: string): string {
  return parseNodeKey(nodeKey).staticPath;
}

/**
 * Test whether a resolved Node Key is below any of the provided anchor keys.
 * Comparison is based on static paths plus ordered dynamic frames: a node is
 * below an anchor when it is not the exact same key, its static path is at or
 * below the anchor's static path, and its dynamic frames stay within the
 * anchor's runtime instance.
 */
export function isNodeKeyBelowAnyAnchor(nodeKey: string, anchorKeys: string[]): boolean {
  if (anchorKeys.length === 0) return false;
  const nodeParsed = parseNodeKey(nodeKey);

  for (const anchorKey of anchorKeys) {
    // Skip if this IS the anchor (not below it)
    if (nodeKey === anchorKey) continue;
    const anchorParsed = parseNodeKey(anchorKey);
    // Check that the node's static path is below the anchor's static path
    if (nodeParsed.staticPath === anchorParsed.staticPath || nodeParsed.staticPath.startsWith(`${anchorParsed.staticPath}/`)) {
      // Verify dynamic scope alignment: the node must be in the same dynamic
      // scope as the anchor (or a nested sub-scope under it)
      if (isDynamicFramePrefix(anchorParsed.dynamicFrames, nodeParsed.dynamicFrames)) {
        return true;
      }
    }
  }
  return false;
}

/**
 * Check whether the anchor's ordered dynamic frames are a prefix of the node's
 * dynamic frames. This preserves subworkflow/fanout nesting boundaries when
 * the same inner item/lane values repeat under different outer instances.
 */
function isDynamicFramePrefix(anchorFrames: NodeKeyDynamic[], nodeFrames: NodeKeyDynamic[]): boolean {
  anchorFrames = collapseAdjacentDuplicateFrames(anchorFrames);
  nodeFrames = collapseAdjacentDuplicateFrames(nodeFrames);
  if (anchorFrames.length > nodeFrames.length) return false;

  return anchorFrames.every((anchorFrame, index) => {
    const nodeFrame = nodeFrames[index]!;
    if (anchorFrame.fanoutItemId !== undefined && nodeFrame.fanoutItemId !== anchorFrame.fanoutItemId) {
      return false;
    }
    if (anchorFrame.laneId !== undefined && nodeFrame.laneId !== anchorFrame.laneId) {
      return false;
    }
    if (
      anchorFrame.parallelBranchId !== undefined &&
      !isParallelBranchInScope(nodeFrame.parallelBranchId, anchorFrame.parallelBranchId)
    ) {
      return false;
    }
    if (anchorFrame.loopRound !== undefined && nodeFrame.loopRound !== anchorFrame.loopRound) {
      return false;
    }
    return true;
  });
}

function collapseAdjacentDuplicateFrames(frames: NodeKeyDynamic[]): NodeKeyDynamic[] {
  const collapsed: NodeKeyDynamic[] = [];
  for (const frame of frames) {
    const previous = collapsed.at(-1);
    if (previous !== undefined && isSameDynamicFrame(previous, frame)) continue;
    collapsed.push(frame);
  }
  return collapsed;
}

function isSameDynamicFrame(left: NodeKeyDynamic, right: NodeKeyDynamic): boolean {
  return (
    left.fanoutItemId === right.fanoutItemId &&
    left.laneId === right.laneId &&
    left.parallelBranchId === right.parallelBranchId &&
    left.loopRound === right.loopRound
  );
}

/**
 * Test whether a resolved Node Key addresses `staticPath` or a descendant of it.
 * The comparison ignores dynamic dimensions.
 */
export function isNodeKeyAtOrBelow(nodeKey: string, staticPath: string): boolean {
  const nodeStaticPath = staticNodePathFromKey(nodeKey);
  return nodeStaticPath === staticPath || nodeStaticPath.startsWith(`${staticPath}/`);
}

/**
 * Test whether a resolved Node Key belongs to the provided dynamic scope.
 * Required scope dimensions are encoded with the same rules as Node Key
 * resolution before comparison.
 */
export function isNodeKeyInDynamicScope(nodeKey: string, dynamic: NodeKeyDynamic): boolean {
  if (isEmptyDynamicScope(dynamic)) {
    return true;
  }

  const requiredFrames = dynamicFrames(dynamic);
  return isDynamicFrameSubsequence(requiredFrames, collectDynamicFrames(nodeKey));
}

function isDynamicFrameSubsequence(requiredFrames: NodeKeyDynamic[], nodeFrames: NodeKeyDynamic[]): boolean {
  let nodeIndex = 0;
  for (const required of requiredFrames) {
    let matched = false;
    for (; nodeIndex < nodeFrames.length; nodeIndex++) {
      if (isDynamicFrameScopeMatch(nodeFrames[nodeIndex]!, required)) {
        matched = true;
        nodeIndex++;
        break;
      }
    }
    if (!matched) return false;
  }
  return true;
}

function isDynamicFrameScopeMatch(frame: DynamicFrame, required: NodeKeyDynamic): boolean {
  return (
    (required.fanoutItemId === undefined || frame.fanoutItemId === required.fanoutItemId) &&
    (required.laneId === undefined || frame.laneId === required.laneId) &&
    (required.parallelBranchId === undefined || isParallelBranchInScope(frame.parallelBranchId, required.parallelBranchId)) &&
    (required.loopRound === undefined || frame.loopRound === required.loopRound)
  );
}

function collectDynamicFrames(nodeKey: string): DynamicFrame[] {
  const frames: DynamicFrame[] = [];
  let current: DynamicFrame = {};

  for (const segment of nodeKey.split("/")) {
    if (!DYNAMIC_SEGMENT.test(segment)) {
      pushDynamicFrame(frames, current);
      current = {};
      continue;
    }

    const [kind, ...rest] = segment.split(":");
    const value = rest.join(":");

    // item starts a fanout frame; branch/round start their own composite frames.
    if ((kind === "item" || kind === "branch" || kind === "round") && !isEmptyDynamicScope(current)) {
      pushDynamicFrame(frames, current);
      current = {};
    }

    switch (kind) {
      case "item":
        current.fanoutItemId = value;
        break;
      case "lane":
        current.laneId = value;
        break;
      case "branch":
        current.parallelBranchId = value;
        break;
      case "round":
        current.loopRound = Number(value);
        break;
    }
  }

  pushDynamicFrame(frames, current);
  return frames;
}

function dynamicFrames(dynamic: NodeKeyDynamic): NodeKeyDynamic[] {
  if (dynamic.frames && dynamic.frames.length > 0) return dynamic.frames;
  const { frames: _frames, ...frame } = dynamic;
  return isEmptyDynamicScope(frame) ? [] : splitDynamicFrame(frame);
}

function splitDynamicFrame(frame: NodeKeyDynamic): NodeKeyDynamic[] {
  const frames: NodeKeyDynamic[] = [];
  if (frame.fanoutItemId !== undefined || frame.laneId !== undefined) {
    frames.push({
      fanoutItemId: frame.fanoutItemId === undefined ? undefined : sanitizeValue(String(frame.fanoutItemId)),
      laneId: frame.laneId === undefined ? undefined : sanitizeValue(String(frame.laneId))
    });
  }
  if (frame.parallelBranchId !== undefined) frames.push({ parallelBranchId: sanitizeValue(String(frame.parallelBranchId)) });
  if (frame.loopRound !== undefined) frames.push({ loopRound: frame.loopRound });
  return frames;
}

function pushDynamicFrame(frames: DynamicFrame[], frame: DynamicFrame): void {
  if (!isEmptyDynamicScope(frame)) {
    frames.push(frame);
  }
}

function isEmptyDynamicScope(dynamic: NodeKeyDynamic): boolean {
  return (
    dynamic.fanoutItemId === undefined &&
    dynamic.laneId === undefined &&
    dynamic.parallelBranchId === undefined &&
    dynamic.loopRound === undefined
  );
}

function isParallelBranchInScope(parsedBranch: string | undefined, scopeBranch: string): boolean {
  const sanitizedScope = sanitizeValue(String(scopeBranch));
  return parsedBranch === sanitizedScope || parsedBranch?.startsWith(`${sanitizedScope}.`) === true;
}

/** Prefix a resolved child Node Key with a parent Node Key, preserving nesting. */
export function withNodeKeyPrefix(prefix: string | undefined, nodeKey: string): string {
  return prefix ? `${prefix}/${nodeKey}` : nodeKey;
}

/** Sanitize a dynamic value to be filesystem-safe (shared with key-scope matching). */
function sanitizeValue(value: string): string {
  // Replace filesystem-unsafe characters with underscores
  return value.replace(/[/\\:*?"<>|]/g, "_");
}

/**
 * Encode a node key for use as a filesystem filename.
 * Replaces "/" with ":" to flatten the hierarchical key into a flat name,
 * then appends ".json".
 * This encoding must be used consistently across store.ts and artifacts.ts.
 */
export function encodeNodeKeyForFs(nodeKey: string): string {
  return encodeNodeKeyForDir(nodeKey) + ".json";
}

/**
 * Encode a node key for use as a filesystem directory name.
 * Replaces "/" with ":" to flatten the hierarchical key into a flat name.
 * Use this when you need the encoded key without a file extension
 * (e.g., artifact directories, path segments).
 */
export function encodeNodeKeyForDir(nodeKey: string): string {
  return nodeKey.replace(/\//g, ":");
}

File: crates/acpus-runtime/src/keys.rs
use crate::NodeKeyDynamic;
use acpus_core::NodeKeyTemplate;
use serde::{Deserialize, Serialize};
use sha2::{Digest, Sha256};

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
pub struct ParsedNodeKey {
    #[serde(rename = "nodeKey")]
    pub node_key: String,
    #[serde(rename = "staticPath")]
    pub static_path: String,
    #[serde(rename = "staticSegments")]
    pub static_segments: Vec<String>,
    pub dynamic: NodeKeyDynamic,
    #[serde(rename = "dynamicFrames")]
    pub dynamic_frames: Vec<NodeKeyDynamic>,
}

pub fn resolve_node_key(template: &NodeKeyTemplate, dynamic: &NodeKeyDynamic) -> String {
    let mut parts = vec![template.node_path.clone()];
    let frames = if dynamic.frames.is_empty() {
        vec![dynamic.clone()]
    } else {
        dynamic.frames.clone()
    };
    for frame in frames {
        if let Some(value) = frame.fanout_item_id {
            parts.push(format!("item:{}", sanitize_value(&value)));
        }
        if let Some(value) = frame.lane_id {
            parts.push(format!("lane:{}", sanitize_value(&value)));
        }
        if let Some(value) = frame.parallel_branch_id {
            parts.push(format!("branch:{}", sanitize_value(&value)));
        }
        if let Some(value) = frame.loop_round {
            parts.push(format!("round:{value}"));
        }
    }
    parts.join("/")
}

pub fn with_node_key_prefix(prefix: Option<&str>, node_key: &str) -> String {
    prefix
        .map(|prefix| format!("{prefix}/{node_key}"))
        .unwrap_or_else(|| node_key.to_string())
}

pub fn append_dynamic_frame(dynamic: &NodeKeyDynamic, frame: NodeKeyDynamic) -> NodeKeyDynamic {
    let mut frames = runtime_dynamic_frames(dynamic);
    frames.push(clear_nested_frames(frame));
    collapse_dynamic_frames(frames)
}

pub fn append_dynamic_frames(dynamic: &NodeKeyDynamic, child: &NodeKeyDynamic) -> NodeKeyDynamic {
    let mut frames = runtime_dynamic_frames(dynamic);
    frames.extend(runtime_dynamic_frames(child));
    collapse_dynamic_frames(frames)
}

pub fn replace_current_dynamic_frame(
    dynamic: &NodeKeyDynamic,
    frame: NodeKeyDynamic,
) -> NodeKeyDynamic {
    let mut frames = runtime_dynamic_frames(dynamic);
    if frames.is_empty() {
        frames.push(clear_nested_frames(frame));
    } else {
        let last = frames.len() - 1;
        frames[last] = clear_nested_frames(frame);
    }
    collapse_dynamic_frames(frames)
}

pub fn current_dynamic_frame(dynamic: &NodeKeyDynamic) -> NodeKeyDynamic {
    runtime_dynamic_frames(dynamic)
        .last()
        .cloned()
        .unwrap_or_default()
}

fn clear_nested_frames(mut frame: NodeKeyDynamic) -> NodeKeyDynamic {
    frame.frames.clear();
    frame
}

pub fn nested_parallel_branch_dynamic(dynamic: &NodeKeyDynamic, branch_id: &str) -> NodeKeyDynamic {
    let mut current = current_dynamic_frame(dynamic);
    current.parallel_branch_id = Some(match current.parallel_branch_id {
        Some(parent) => format!("{parent}.{branch_id}"),
        None => branch_id.to_string(),
    });
    replace_current_dynamic_frame(dynamic, current)
}

pub fn parse_node_key(node_key: &str) -> ParsedNodeKey {
    let mut static_segments = Vec::new();
    let mut dynamic_segments = Vec::new();
    for segment in node_key.split('/') {
        if matches!(
            segment.split_once(':'),
            Some(("item" | "lane" | "round" | "branch", _))
        ) {
            dynamic_segments.push(segment.to_string());
        } else {
            static_segments.push(segment.to_string());
        }
    }
    let mut dynamic = NodeKeyDynamic::default();
    for segment in &dynamic_segments {
        set_dynamic(&mut dynamic, segment);
    }
    let dynamic_frames = build_dynamic_frames(&dynamic_segments);
    ParsedNodeKey {
        node_key: node_key.to_string(),
        static_path: static_segments.join("/"),
        static_segments,
        dynamic,
        dynamic_frames,
    }
}

pub fn static_node_path_from_key(node_key: &str) -> String {
    parse_node_key(node_key).static_path
}

pub fn is_node_key_at_or_below(node_key: &str, static_path: &str) -> bool {
    let node_static_path = static_node_path_from_key(node_key);
    node_static_path == static_path || node_static_path.starts_with(&format!("{static_path}/"))
}

pub fn is_node_key_below_any_anchor(node_key: &str, anchors: &[String]) -> bool {
    if anchors.is_empty() {
        return false;
    }
    let parsed = parse_node_key(node_key);
    anchors.iter().any(|anchor| {
        if node_key == anchor {
            return false;
        }
        let anchor = parse_node_key(anchor);
        (parsed.static_path == anchor.static_path
            || parsed
                .static_path
                .starts_with(&format!("{}/", anchor.static_path)))
            && is_dynamic_frame_prefix(&anchor.dynamic_frames, &parsed.dynamic_frames)
    })
}

pub fn is_node_key_in_dynamic_scope(node_key: &str, dynamic: &NodeKeyDynamic) -> bool {
    if is_empty_dynamic_scope(dynamic) {
        return true;
    }
    is_dynamic_frame_subsequence(
        &scope_dynamic_frames(dynamic),
        &collect_dynamic_frames(node_key),
    )
}

pub fn storage_key(node_key: &str) -> String {
    let mut slug = String::new();
    let mut last_dash = false;
    for c in node_key.chars() {
        if c == '-' {
            if !last_dash {
                slug.push('-');
                last_dash = true;
            }
        } else if c.is_ascii_alphanumeric() || matches!(c, '.' | '_') {
            slug.push(c);
            last_dash = false;
        } else if !last_dash {
            slug.push('-');
            last_dash = true;
        }
    }
    let slug = slug.trim_matches('-');
    let slug = if slug.is_empty() || matches!(slug, "." | "..") {
        "node"
    } else {
        slug
    };
    let hash = hex::encode(Sha256::digest(node_key.as_bytes()));
    format!("{}--{}", shorten_storage_slug(slug), &hash[..16])
}

fn sanitize_value(value: &str) -> String {
    value
        .chars()
        .map(|c| {
            if matches!(c, '/' | '\\' | ':' | '*' | '?' | '"' | '<' | '>' | '|') {
                '_'
            } else {
                c
            }
        })
        .collect()
}

fn shorten_storage_slug(slug: &str) -> String {
    const SLUG_LENGTH: usize = 70;
    if slug.len() <= SLUG_LENGTH {
        return slug.to_string();
    }
    let tail_length = (SLUG_LENGTH - 3) / 2;
    let head_length = SLUG_LENGTH - 3 - tail_length;
    format!(
        "{}...{}",
        &slug[..head_length],
        &slug[slug.len() - tail_length..]
    )
}

fn runtime_dynamic_frames(dynamic: &NodeKeyDynamic) -> Vec<NodeKeyDynamic> {
    if !dynamic.frames.is_empty() {
        return dynamic.frames.clone();
    }
    let frame = clear_nested_frames(dynamic.clone());
    if frame == NodeKeyDynamic::default() {
        Vec::new()
    } else {
        vec![frame]
    }
}

fn scope_dynamic_frames(dynamic: &NodeKeyDynamic) -> Vec<NodeKeyDynamic> {
    if !dynamic.frames.is_empty() {
        return dynamic.frames.clone();
    }
    split_dynamic_frame(&clear_nested_frames(dynamic.clone()))
}

fn split_dynamic_frame(frame: &NodeKeyDynamic) -> Vec<NodeKeyDynamic> {
    let mut frames = Vec::new();
    if frame.fanout_item_id.is_some() || frame.lane_id.is_some() {
        frames.push(NodeKeyDynamic {
            fanout_item_id: frame.fanout_item_id.as_deref().map(sanitize_value),
            lane_id: frame.lane_id.as_deref().map(sanitize_value),
            ..Default::default()
        });
    }
    if let Some(branch) = &frame.parallel_branch_id {
        frames.push(NodeKeyDynamic {
            parallel_branch_id: Some(sanitize_value(branch)),
            ..Default::default()
        });
    }
    if let Some(loop_round) = frame.loop_round {
        frames.push(NodeKeyDynamic {
            loop_round: Some(loop_round),
            ..Default::default()
        });
    }
    frames
}

fn collapse_dynamic_frames(frames: Vec<NodeKeyDynamic>) -> NodeKeyDynamic {
    let mut dynamic = NodeKeyDynamic {
        frames,
        ..Default::default()
    };
    for frame in &dynamic.frames {
        if let Some(value) = frame.loop_round {
            dynamic.loop_round = Some(value);
        }
        if let Some(value) = &frame.fanout_item_id {
            dynamic.fanout_item_id = Some(value.clone());
        }
        if let Some(value) = &frame.lane_id {
            dynamic.lane_id = Some(value.clone());
        }
        if let Some(value) = &frame.parallel_branch_id {
            dynamic.parallel_branch_id = Some(value.clone());
        }
    }
    dynamic
}

fn collapse_adjacent_duplicate_frames(frames: &[NodeKeyDynamic]) -> Vec<NodeKeyDynamic> {
    let mut collapsed = Vec::new();
    for frame in frames {
        if collapsed.last() == Some(frame) {
            continue;
        }
        collapsed.push(frame.clone());
    }
    collapsed
}

fn is_dynamic_frame_prefix(anchor: &[NodeKeyDynamic], node: &[NodeKeyDynamic]) -> bool {
    let anchor = collapse_adjacent_duplicate_frames(anchor);
    let node = collapse_adjacent_duplicate_frames(node);
    anchor.len() <= node.len()
        && anchor
            .iter()
            .zip(node.iter())
            .all(|(anchor, node)| dynamic_frame_matches(anchor, node))
}

fn is_dynamic_frame_subsequence(required: &[NodeKeyDynamic], node: &[NodeKeyDynamic]) -> bool {
    let mut node_index = 0;
    for required in required {
        let mut matched = false;
        while node_index < node.len() {
            if dynamic_frame_matches(required, &node[node_index]) {
                matched = true;
                node_index += 1;
                break;
            }
            node_index += 1;
        }
        if !matched {
            return false;
        }
    }
    true
}

fn dynamic_frame_matches(anchor: &NodeKeyDynamic, node: &NodeKeyDynamic) -> bool {
    (anchor.fanout_item_id.is_none() || anchor.fanout_item_id == node.fanout_item_id)
        && (anchor.lane_id.is_none() || anchor.lane_id == node.lane_id)
        && anchor.parallel_branch_id.as_ref().is_none_or(|branch| {
            is_parallel_branch_in_scope(node.parallel_branch_id.as_deref(), branch)
        })
        && (anchor.loop_round.is_none() || anchor.loop_round == node.loop_round)
}

fn is_parallel_branch_in_scope(parsed_branch: Option<&str>, scope_branch: &str) -> bool {
    let scope = sanitize_value(scope_branch);
    parsed_branch.is_some_and(|parsed| parsed == scope || parsed.starts_with(&format!("{scope}.")))
}

fn collect_dynamic_frames(node_key: &str) -> Vec<NodeKeyDynamic> {
    let mut frames = Vec::new();
    let mut current = NodeKeyDynamic::default();
    for segment in node_key.split('/') {
        if !is_dynamic_segment(segment) {
            push_dynamic_frame(&mut frames, &mut current);
            continue;
        }
        let Some((kind, _)) = segment.split_once(':') else {
            continue;
        };
        if matches!(kind, "item" | "branch" | "round") && !is_empty_dynamic_scope(&current) {
            push_dynamic_frame(&mut frames, &mut current);
        }
        set_dynamic(&mut current, segment);
    }
    push_dynamic_frame(&mut frames, &mut current);
    frames
}

fn push_dynamic_frame(frames: &mut Vec<NodeKeyDynamic>, current: &mut NodeKeyDynamic) {
    if !is_empty_dynamic_scope(current) {
        frames.push(std::mem::take(current));
    }
}

fn is_empty_dynamic_scope(dynamic: &NodeKeyDynamic) -> bool {
    dynamic.loop_round.is_none()
        && dynamic.fanout_item_id.is_none()
        && dynamic.lane_id.is_none()
        && dynamic.parallel_branch_id.is_none()
}

fn is_dynamic_segment(segment: &str) -> bool {
    matches!(
        segment.split_once(':'),
        Some(("item" | "lane" | "round" | "branch", _))
    )
}

fn set_dynamic(dynamic: &mut NodeKeyDynamic, segment: &str) {
    let Some((kind, value)) = segment.split_once(':') else {
        return;
    };
    match kind {
        "item" => dynamic.fanout_item_id = Some(value.to_string()),
        "lane" => dynamic.lane_id = Some(value.to_string()),
        "branch" => dynamic.parallel_branch_id = Some(value.to_string()),
        "round" => dynamic.loop_round = value.parse().ok(),
        _ => {}
    }
}

fn build_dynamic_frames(segments: &[String]) -> Vec<NodeKeyDynamic> {
    let mut frames = Vec::new();
    let mut current = NodeKeyDynamic::default();
    for segment in segments {
        let begins_scope = matches!(
            segment.split_once(':'),
            Some(("item" | "branch" | "round", _))
        );
        if begins_scope && current != NodeKeyDynamic::default() {
            frames.push(current);
            current = NodeKeyDynamic::default();
        }
        set_dynamic(&mut current, segment);
    }
    if current != NodeKeyDynamic::default() {
        frames.push(current);
    }
    frames
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn key_round_trips_dynamic_parts() {
        let template = NodeKeyTemplate {
            ast_version: 1,
            node_path: "workflow/map/do".into(),
            loop_round: true,
            fanout_item_id: true,
            parallel_branch_id: true,
            lane_id: true,
        };
        let key = resolve_node_key(
            &template,
            &NodeKeyDynamic {
                fanout_item_id: Some("file-a".into()),
                lane_id: Some("0".into()),
                ..Default::default()
            },
        );
        assert_eq!(
            parse_node_key(&key).dynamic.fanout_item_id.unwrap(),
            "file-a"
        );
    }

    #[test]
    fn resolve_node_key_sanitizes_dynamic_values_with_underscores() {
        let template = NodeKeyTemplate {
            ast_version: 1,
            node_path: "workflow/mapped".into(),
            loop_round: false,
            fanout_item_id: true,
            parallel_branch_id: false,
            lane_id: false,
        };

        assert_eq!(
            resolve_node_key(
                &template,
                &NodeKeyDynamic {
                    fanout_item_id: Some("path/to/file".into()),
                    ..Default::default()
                },
            ),
            "workflow/mapped/item:path_to_file"
        );
    }

    #[test]
    fn dynamic_frame_helpers_preserve_boundaries_and_collapsed_fields() {
        let dynamic = append_dynamic_frame(
            &NodeKeyDynamic {
                fanout_item_id: Some("outer".into()),
                lane_id: Some("0".into()),
                ..Default::default()
            },
            NodeKeyDynamic {
                loop_round: Some(1),
                ..Default::default()
            },
        );
        assert_eq!(dynamic.fanout_item_id.as_deref(), Some("outer"));
        assert_eq!(dynamic.lane_id.as_deref(), Some("0"));
        assert_eq!(dynamic.loop_round, Some(1));
        assert_eq!(dynamic.frames.len(), 2);

        let child = NodeKeyDynamic {
            frames: vec![
                NodeKeyDynamic {
                    parallel_branch_id: Some("left".into()),
                    ..Default::default()
                },
                NodeKeyDynamic {
                    loop_round: Some(2),
                    ..Default::default()
                },
            ],
            parallel_branch_id: Some("left".into()),
            loop_round: Some(2),
            ..Default::default()
        };
        let appended = append_dynamic_frames(
            &NodeKeyDynamic {
                fanout_item_id: Some("outer".into()),
                lane_id: Some("0".into()),
                ..Default::default()
            },
            &child,
        );
        assert_eq!(appended.fanout_item_id.as_deref(), Some("outer"));
        assert_eq!(appended.parallel_branch_id.as_deref(), Some("left"));
        assert_eq!(appended.loop_round, Some(2));
        assert_eq!(appended.frames.len(), 3);

        let replaced = replace_current_dynamic_frame(
            &dynamic,
            NodeKeyDynamic {
                loop_round: Some(2),
                ..Default::default()
            },
        );
        assert_eq!(replaced.loop_round, Some(2));
        assert_eq!(replaced.frames.len(), 2);
        assert_eq!(
            current_dynamic_frame(&NodeKeyDynamic {
                fanout_item_id: Some("outer".into()),
                lane_id: Some("0".into()),
                ..Default::default()
            })
            .fanout_item_id
            .as_deref(),
            Some("outer")
        );
    }

    #[test]
    fn node_key_below_anchor_uses_static_path_and_dynamic_frames() {
        assert!(is_node_key_below_any_anchor(
            "workflow/step-a/child/item:x/lane:0",
            &["workflow/step-a/item:x/lane:0".to_string()]
        ));
        assert!(!is_node_key_below_any_anchor(
            "workflow/step-a/item:x/lane:0",
            &["workflow/step-a/item:x/lane:0".to_string()]
        ));
        assert!(!is_node_key_below_any_anchor(
            "workflow/mapped/work/item:lane-b/lane:0",
            &["workflow/mapped/work/item:lane-a/lane:0".to_string()]
        ));

        let anchor = "workflow/mapped/sub/item:a/lane:0/workflow/child_mapped/item:a/lane:0";
        assert!(!is_node_key_below_any_anchor(
            "workflow/mapped/sub/item:b/lane:1/workflow/child_mapped/child_work/item:x/lane:0",
            &[anchor.to_string()]
        ));
        assert!(is_node_key_below_any_anchor(
            "workflow/mapped/sub/item:a/lane:0/workflow/child_mapped/child_work/item:x/lane:0",
            &[anchor.to_string()]
        ));
    }

    #[test]
    fn node_key_at_or_below_ignores_dynamic_dimensions() {
        assert!(is_node_key_at_or_below("workflow/build", "workflow/build"));
        assert!(is_node_key_at_or_below(
            "workflow/aggregate/tally/round:1",
            "workflow/aggregate"
        ));
        assert!(!is_node_key_at_or_below(
            "workflow/publish",
            "workflow/build"
        ));
    }

    #[test]
    fn node_key_dynamic_scope_matches_repeated_frames_and_branch_prefixes() {
        let node_key = "workflow/outer/item:outer/lane:0/workflow/child/item:inner/lane:0";
        assert!(is_node_key_in_dynamic_scope(
            node_key,
            &NodeKeyDynamic {
                fanout_item_id: Some("outer".into()),
                lane_id: Some("0".into()),
                ..Default::default()
            }
        ));
        assert!(is_node_key_in_dynamic_scope(
            node_key,
            &NodeKeyDynamic {
                fanout_item_id: Some("inner".into()),
                lane_id: Some("0".into()),
                ..Default::default()
            }
        ));
        assert!(!is_node_key_in_dynamic_scope(
            node_key,
            &NodeKeyDynamic {
                fanout_item_id: Some("inner".into()),
                lane_id: Some("1".into()),
                ..Default::default()
            }
        ));
        assert!(is_node_key_in_dynamic_scope(
            "workflow/x/branch:0.1.2",
            &NodeKeyDynamic {
                parallel_branch_id: Some("0.1".into()),
                ..Default::default()
            }
        ));
        assert!(!is_node_key_in_dynamic_scope(
            "workflow/x/branch:01",
            &NodeKeyDynamic {
                parallel_branch_id: Some("0".into()),
                ..Default::default()
            }
        ));
        assert!(is_node_key_in_dynamic_scope(
            "workflow/mapped/item:path_to_file",
            &NodeKeyDynamic {
                fanout_item_id: Some("path/to/file".into()),
                ..Default::default()
            }
        ));
    }

    #[test]
    fn storage_key_matches_bounded_typescript_shape() {
        let key = "workflow/mapped/item:file-a/lane:0";
        let storage = storage_key(key);

        assert!(storage.starts_with("workflow-mapped-item-file-a-lane-0--"));
        assert_eq!(
            storage.len(),
            "workflow-mapped-item-file-a-lane-0--".len() + 16
        );
        assert!(storage_key("////::::").starts_with("node--"));

        let long = storage_key(&format!("workflow/{}/final-step", "middle-".repeat(80)));
        let slug = long.split_once("--").unwrap().0;
        assert!(slug.len() <= 70);
        assert!(slug.starts_with("workflow-middle-"));
        assert!(slug.contains("..."));
        assert!(slug.ends_with("middle-final-step"));
    }

    #[test]
    fn nested_parallel_branch_replaces_current_frame_and_keeps_ancestry() {
        let dynamic = append_dynamic_frame(
            &NodeKeyDynamic::default(),
            NodeKeyDynamic {
                fanout_item_id: Some("file-a".into()),
                lane_id: Some("0".into()),
                ..Default::default()
            },
        );
        let dynamic = nested_parallel_branch_dynamic(&dynamic, "left");
        let dynamic = nested_parallel_branch_dynamic(&dynamic, "inner");
        let template = NodeKeyTemplate {
            ast_version: 1,
            node_path: "workflow/fanout/$do/par/$left/step".into(),
            loop_round: true,
            fanout_item_id: true,
            parallel_branch_id: true,
            lane_id: true,
        };

        assert_eq!(
            resolve_node_key(&template, &dynamic),
            "workflow/fanout/$do/par/$left/step/item:file-a/lane:0/branch:left.inner"
        );
        assert_eq!(dynamic.parallel_branch_id.as_deref(), Some("left.inner"));
        assert_eq!(
            dynamic.frames[0].parallel_branch_id.as_deref(),
            Some("left.inner")
        );
    }
}
